lcls_tools/common/data_analysis/bmad_modeling/bmad_modeling.py

```diff
@@ -0,0 +1,387 @@
+    pvlist = set()
+    for dm_key, map in all_data_maps.items():
+        if dm_key in ["cavities", "quad"]:
+            elements = (map.data["bmad_name"].to_list())
+            pvs = map.pvlist
+            model_elements = []
+            full_model_elements = tao.lat_list("*", "ele.name")
+            for mdl_ele in full_model_elements:
+                if "#2" in mdl_ele:  # Deal with split elements in bmad
+                    continue
+                new_ele = mdl_ele.replace("#1", "")
+                model_elements.append(new_ele)
+            for indx, ele in enumerate(elements):
+                if ele in model_elements:
+                    pvlist.add(pvs[indx])
+        if dm_key.startswith('K'):  # One for each Klystron station
+            for pv in map.pvlist:
+                if "BEAMCODE" in pv:
+                    pv = map.accelerate_pvname
+                pvlist.add(pv)
+    return list(pvlist)
+
+
+def get_energy_gain_pvlist(beam_path):
+    """Interim function to get list of EDES PVs from YAML file"""
+    pvlist = []
+    yaml_dir = '/sdf/home/c/colocho/lcls-tools/lcls_tools/' \
+               'common/data_analysis/bmad_modeling/yaml/'
+    with open(yaml_dir + 'energy_measurements.yml', 'r') as file:
+        engy_meas = yaml.safe_load(file)[beam_path[0:2]]
+    [pvlist.append(engy_meas[key]) for key in engy_meas.keys()]
+    return pvlist
+
+
+def evaluate_tao(tao, tao_cmds):
+    """
+    Evaluate set of tao commands, toggles lattice_calculation OFF/ON
+    in between command list
+    """
+    tao.cmd("set global lattice_calc_on = F")
+    tao.cmds(tao_cmds)
+    tao.cmd("set global lattice_calc_on = T")
+    output = get_output(tao)
+    return output
+
+
+def get_twiss(tao, element, which="design"):
+    """Gets twiss for element, which can be model, desing or base"""
+    result = tao.ele_twiss(element, which=which)
+    return [result[p] for p in ["beta_a", "alpha_a", "beta_b", "alpha_b"]]
+
+
+def list_to_numpy_array(x, val_type='Ampl', ampl_reject=0.5, none_value=0):
+    """Convert None to 0.0 for values in x, if is_phase convert to radians
+    if is_ampl set values less than 0.5 Mev to zero"""
+    x_np = np.array(x)
+    x_np = np.where(x is None, np.nan, x).astype(float)
+    x_np = np.nan_to_num(x_np, nan=0.0)
+    if val_type == 'Ampl':
+        x_np = np.where(x_np < ampl_reject, 0, x_np)
+    if val_type == 'Phas':
+        x_np = np.deg2rad(x_np)
+    return x_np
+
+
+def bmag(twiss, twiss_reference):
+    """Calculates BMAG from imput twiss and reference twiss"""
+    beta_a, alpha_a, beta_b, alpha_b = twiss
+    beta_a_ref, alpha_a_ref, beta_b_ref, alpha_b_ref = twiss_reference
+    bmag_a = bmag_func(beta_a, alpha_a, beta_a_ref, alpha_a_ref)
+    bmag_b = bmag_func(beta_b, alpha_b, beta_b_ref, alpha_b_ref)
+    return (bmag_a, bmag_b)
+
+
+def bmag_func(bb, ab, bl, al):
+    """bmag function"""
+    return 1 / 2 * (bl / bb + bb / bl + bb * bl * (ab / bb - al / bl) ** 2)
```

Comment:
can you add a comment on what the inputs are?

Reply:
o.k.